### PR TITLE
Move module loop to setup

### PR DESCRIPTION
**Pull request by Jingyan95**
Moving expensive (module) loops to Setup.cc which is called at BeginRun.

### CODE REVIEW

File: L1Trigger/TrackTrigger/python/ProducerHPH_cfi.py [hunk 1]
```diff
@@ -1,11 +1,15 @@
...
 # ParameterSet used by HitPatternHelper
 
 HitPatternHelper_params = cms.PSet (
 
   hphDebug   = cms.bool(False),   
   useNewKF   = cms.bool(False),
-  deltaTanL  = cms.double(0.125)
-
+  oldKFPSet  = cms.PSet(TMTrackProducer_params.EtaSectors),
+  deltaTanL  = cms.double(0.125),
```
Comment (tomalin) [inline L1Trigger/TrackTrigger/python/ProducerHPH_cfi.py]: Add comments explaining what these 3 parameters are and how the numbers are chosen. Also, I think deltaTanL is only used in your code if the NewKF is being used?

File: L1Trigger/TrackTrigger/src/HitPatternHelper.cc [hunk 2]
```diff
@@ -15,11 +15,59 @@ namespace hph {
...
+        double z0Digi_ = (2 * beamWindowZ_ / z0Nbins_) * (z0Bin + 1) - beamWindowZ_;
+        std::vector<tt::SensorModule>& layers_ = layerEncoding_[cotBin][z0Bin];
+        //Looping over sensor modules to make predictions on which layers particles are expected to hit
+        for (const tt::SensorModule& sm : setupTT_.sensorModules()) {
+          double d = (z0Digi_ - sm.z() + sm.r() * cotDigi_) / (sm.cosTilt() - sm.sinTilt() * cotDigi_);
+          double d_p = (z0Digi_ - sm.z() + sm.r() * (cotDigi_ + deltaTanL_ / 2)) /
+                       (sm.cosTilt() - sm.sinTilt() * (cotDigi_ + deltaTanL_ / 2));
+          double d_m = (z0Digi_ - sm.z() + sm.r() * (cotDigi_ - deltaTanL_ / 2)) /
+                       (sm.cosTilt() - sm.sinTilt() * (cotDigi_ - deltaTanL_ / 2));
+          if (useNewKF_ &&
```
Comment (tomalin) [inline L1Trigger/TrackTrigger/src/HitPatternHelper.cc]: Why is the logic here different for assigning a SensorModule to an (eta,z0) sector depending if NewKF is used or not?

File: L1Trigger/TrackTrigger/src/HitPatternHelper.cc [hunk 3]
```diff
@@ -54,13 +101,15 @@ namespace hph {
         cot_(cot),
         z0_(z0),
         setup_(setup),
+        layers_(setup_->layerEncoding(setup_->digiCot(cot_), setup_->digiZ0(z0_))),
```
Comment (tomalin) [inline L1Trigger/TrackTrigger/src/HitPatternHelper.cc]: Eliminate magic number 11. Or at least add comment explaining where its value comes from.

**Comment by tomalin** [comment]
Please say in Description if you've checked this reduces CPU use.

**Comment by tomalin** [comment]
I believe the HPH is used only in L1TrackNtupleMaker.cc to calculate variables such as tmp_trk_nPSstub_hitpattern that are added to the TTree? So if HPH contains bugs, they won't show up (except if compilation errors) in the CI test checks of tracking efficiency? Have you therefore checked that these variables look sensible for both NewKF & OldKF?

**Comment by tomalin** [comment]
The Hybrid code crashed in the CI test https://gitlab.cern.ch/cms-l1tk/cmssw_CI/-/pipelines/4155053 .

**Comment by Jingyan95** [comment]
@tomalin I moved the HPH files under TrackFindingTracklet package such that Thomas's functions from TrackerTFP can be included without creating circular dependency.

**Comment by Jingyan95** [comment]
Hi @tomalin , I wanted to check if you had some time to look at this PR ? The primary comment (CPU usage) was addressed by the recent commit and the hitpatternhelper module now uses very minimal CPU.

**Comment by tomalin** [comment]
> @tomalin I moved the HPH files under TrackFindingTracklet package such that Thomas's functions from TrackerTFP can be included without creating circular dependency.

@Jingyan95 Can you please explicitely answer each of the review comments. Whilst some remain unanswered, I assume the work's not yet been done, so don't bother rereviewing.

File: L1Trigger/TrackTrigger/python/ProducerHPH_cfi.py (hunk 1 above)
Comment (Jingyan95) [reply L1Trigger/TrackTrigger/python/ProducerHPH_cfi.py]: Done

File: L1Trigger/TrackTrigger/src/HitPatternHelper.cc (hunk 2 above)
Comment (Jingyan95) [reply L1Trigger/TrackTrigger/src/HitPatternHelper.cc]: Done, the logic is the same for old and new kf now

File: L1Trigger/TrackTrigger/src/HitPatternHelper.cc (hunk 3 above)
Comment (Jingyan95) [reply L1Trigger/TrackTrigger/src/HitPatternHelper.cc]: Done, added new description

**Comment by Jingyan95** [comment]
> Please say in Description if you've checked this reduces CPU use.

CPU usage is negligible now

**Comment by Jingyan95** [comment]
> I believe the HPH is used only in L1TrackNtupleMaker.cc to calculate variables such as tmp_trk_nPSstub_hitpattern that are added to the TTree? So if HPH contains bugs, they won't show up (except if compilation errors) in the CI test checks of tracking efficiency? Have you therefore checked that these variables look sensible for both NewKF & OldKF?

The performance of the HPH has been check and it is consistent with the previous version

**Comment by Jingyan95** [comment]
> > I believe the HPH is used only in L1TrackNtupleMaker.cc to calculate variables such as tmp_trk_nPSstub_hitpattern that are added to the TTree? So if HPH contains bugs, they won't show up (except if compilation errors) in the CI test checks of tracking efficiency? Have you therefore checked that these variables look sensible for both NewKF & OldKF?
> 
> The performance of the HPH has been checked and it is consistent with the previous version
Old KF:
[TTbar_PU200_D76_FPR_hitpattern.pdf](https://github.com/cms-L1TK/cmssw/files/9536653/TTbar_PU200_D76_FPR_hitpattern.pdf)
[TTbar_PU200_D76_TPR_hitpattern.pdf](https://github.com/cms-L1TK/cmssw/files/9536654/TTbar_PU200_D76_TPR_hitpattern.pdf)
New KF:
[TTbar_PU200_D76_NEWKF_FPR_hitpattern.pdf](https://github.com/cms-L1TK/cmssw/files/9536661/TTbar_PU200_D76_NEWKF_FPR_hitpattern.pdf)
[TTbar_PU200_D76_NEWKF_TPR_hitpattern.pdf](https://github.com/cms-L1TK/cmssw/files/9536662/TTbar_PU200_D76_NEWKF_TPR_hitpattern.pdf)

**Comment by skinnari** [comment]
hi @tomalin @Jingyan95 it seems that all comments have been addressed, can this PR be merged? thanks!

**Review by tomalin — APPROVED** [review]
approved

**Pull request merged**